bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyLinesProvider.java

```diff
@@ -0,0 +1,174 @@
+/*******************************************************************************
+ * Copyright (c) 2024 SAP SE.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     SAP SE - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.ui.internal.texteditor.stickyscroll;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.eclipse.swt.custom.StyledText;
+
+import org.eclipse.jface.text.ITextViewerExtension5;
+import org.eclipse.jface.text.source.ISourceViewer;
+
+/**
+ * This class provides sticky lines for the given source code in the source viewer. The
+ * implementation is completely based on indentation and therefore should work by default for
+ * several languages.
+ */
+public class StickyLinesProvider {
```

Comment:
For the Future: how about we extract an IStickyLinesProvider and ask each editor to provide their own sticky lines provider? This fixes my comment about "normal text files" having weird behavior

Reply:
Yes, I also though about that. But I wanted to wait until we have two implementations to have a better feeling what is needed before creating a new API.